### PR TITLE
Add query to check for impossible predicates, use it to skip `NoopMethodCall` and `Inline`

### LINKED ISSUE #93008: ICE with "--crate-type lib --Z mir-opt-level-3": `Unimplemented` selecting `Binder(<&str as Foo>, [])` during codegen
### Code

```Rust
#![feature(trivial_bounds)]
trait Foo {
    fn test(self);
}
fn baz<T>()
where
    &'static str: Foo,
{
    "Foo".test()
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (bd3cb5256 2022-01-16)
binary: rustc
commit-hash: bd3cb52565faab2755ff1bdb54d88bc91f47b4b9
commit-date: 2022-01-16
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 13.0.0
```

### Error output

```
warning: function is never used: `baz`
 --> mutant_7722893.rs:5:4
  |
5 | fn baz<T>()
  |    ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: trait bound &'static str: Foo does not depend on any type or lifetime parameters
 --> mutant_7722893.rs:7:19
  |
7 |     &'static str: Foo,
  |                   ^^^
  |
  = note: `#[warn(trivial_bounds)]` on by default

warning: 2 warnings emitted

error: internal compiler error: Encountered error `Unimplemented` selecting `Binder(<&str as Foo>, [])` during codegen
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/codegen.rs:68:32

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1188:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.60.0-nightly (bd3cb5256 2022-01-16) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=3 --crate-type lib

query stack during panic:
end of query stack
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
warning: function is never used: `baz`
 --> mutant_7722893.rs:5:4
  |
5 | fn baz<T>()
  |    ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: trait bound &'static str: Foo does not depend on any type or lifetime parameters
 --> mutant_7722893.rs:7:19
  |
7 |     &'static str: Foo,
  |                   ^^^
  |
  = note: `#[warn(trivial_bounds)]` on by default

warning: 2 warnings emitted

error: internal compiler error: Encountered error `Unimplemented` selecting `Binder(<&str as Foo>, [])` during codegen
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/codegen.rs:68:32

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1188:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/bd3cb52565faab2755ff1bdb54d88bc91f47b4b9/library/std/src/panicking.rs:577:5
   1: core::panicking::panic_fmt
             at /rustc/bd3cb52565faab2755ff1bdb54d88bc91f47b4b9/library/core/src/panicking.rs:110:14
   2: core::panicking::panic_display::<&str>
   3: <rustc_errors::HandlerInner>::flush_delayed
   4: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
   5: core::ptr::drop_in_place::<rustc_session::parse::ParseSess>
   6: <alloc::rc::Rc<rustc_session::session::Session> as core::ops::drop::Drop>::drop
   7: core::ptr::drop_in_place::<rustc_interface::interface::Compiler>
   8: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
   9: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>
  10: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.60.0-nightly (bd3cb5256 2022-01-16) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=3 --crate-type lib

query stack during panic:
end of query stack
```

</p>
</details>

### LINKED ISSUE #94680: ICE with generics.
### Code

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3f6d8b903da8b01348137330371e7ff5

```Rust
fn main() {
   println!("{:?}", {
   type T = ();

     pub fn cloneit (it: &'_ mut T) -> (&'_ mut T, &'_ mut T)
where
    for<'any> &'any mut T : Clone, {
        (it.clone(), it)
    }
}); }
```

### Meta

`rustc --version --verbose`:
```
1.61.0-nightly (2022-03-05 c274e4969f058b1c6442)
```

### Error output

```
Compiling playground v0.0.1 (/playground)
warning: type alias is never used: `T`
 [--> src/main.rs:3:4
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4c9140e1c57f836e8185361ea26f0541#)  |
3 |    type T = ();
  |    ^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function is never used: `clone`
 [--> src/main.rs:5:13
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4c9140e1c57f836e8185361ea26f0541#)  |
5 |      pub fn clone (it: &'_ mut T) -> (&'_ mut T, &'_ mut T)
  |             ^^^^^

error: internal compiler error: no errors encountered even though `delay_span_bug` issued

error: internal compiler error: Encountered error `Unimplemented` selecting `Binder(<&mut () as std::clone::Clone>, [])` during codegen
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/codegen.rs:69:32

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1266:13
stack backtrace:
   0:     0x7fce63604d2c - std::backtrace_rs::backtrace::libunwind::trace::hd9d6c66ed2138bb0
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7fce63604d2c - std::backtrace_rs::backtrace::trace_unsynchronized::ha4edb04af5a111d2
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fce63604d2c - std::sys_common::backtrace::_print_fmt::h34bfe6ef54352a7f
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7fce63604d2c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h91638cd7890644b7
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7fce6365ebec - core::fmt::write::h407f77bceaf22b75
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/core/src/fmt/mod.rs:1190:17
   5:     0x7fce635f5801 - std::io::Write::write_fmt::h85de75cd3cb86034
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/io/mod.rs:1655:15
   6:     0x7fce63607c65 - std::sys_common::backtrace::_print::h9303e0f315439c4b
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7fce63607c65 - std::sys_common::backtrace::print::h278cbd6c00b20bf7
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7fce63607c65 - std::panicking::default_hook::{{closure}}::h66dc8e0c0481f9ac
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/panicking.rs:295:22
   9:     0x7fce63607908 - std::panicking::default_hook::h3732f22170eb6fcb
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/panicking.rs:314:9
  10:     0x7fce63d9d4a1 - rustc_driver[a243c3d9738e9cda]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7fce63608501 - std::panicking::rust_panic_with_hook::hfa6a9afb1a6b2eff
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/panicking.rs:702:17
  12:     0x7fce64e2dfb1 - std[54a224735ab911a]::panicking::begin_panic::<rustc_errors[e9d593471e7e5a86]::ExplicitBug>::{closure#0}
  13:     0x7fce64e2dea6 - std[54a224735ab911a]::sys_common::backtrace::__rust_end_short_backtrace::<std[54a224735ab911a]::panicking::begin_panic<rustc_errors[e9d593471e7e5a86]::ExplicitBug>::{closure#0}, !>
  14:     0x7fce64e348ef - std[54a224735ab911a]::panicking::begin_panic::<rustc_errors[e9d593471e7e5a86]::ExplicitBug>
  15:     0x7fce64e41756 - std[54a224735ab911a]::panic::panic_any::<rustc_errors[e9d593471e7e5a86]::ExplicitBug>
  16:     0x7fce664bd16e - <rustc_errors[e9d593471e7e5a86]::HandlerInner as core[f66f8263def5172a]::ops::drop::Drop>::drop
  17:     0x7fce65bfb338 - core[f66f8263def5172a]::ptr::drop_in_place::<rustc_session[7e14f15cd8696afc]::parse::ParseSess>
  18:     0x7fce65bfde8a - <alloc[16beeacba2f7b906]::rc::Rc<rustc_session[7e14f15cd8696afc]::session::Session> as core[f66f8263def5172a]::ops::drop::Drop>::drop
  19:     0x7fce65be652d - core[f66f8263def5172a]::ptr::drop_in_place::<rustc_interface[ee304f5e577cf6b9]::interface::Compiler>
  20:     0x7fce65be6033 - rustc_span[1b916fa4cb305af0]::with_source_map::<core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>, rustc_interface[ee304f5e577cf6b9]::interface::create_compiler_and_run<core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>, rustc_driver[a243c3d9738e9cda]::run_compiler::{closure#1}>::{closure#1}>
  21:     0x7fce65bd3f24 - rustc_interface[ee304f5e577cf6b9]::interface::create_compiler_and_run::<core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>, rustc_driver[a243c3d9738e9cda]::run_compiler::{closure#1}>
  22:     0x7fce65bd1f42 - <scoped_tls[a76d38c46f15ed9e]::ScopedKey<rustc_span[1b916fa4cb305af0]::SessionGlobals>>::set::<rustc_interface[ee304f5e577cf6b9]::interface::run_compiler<core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>, rustc_driver[a243c3d9738e9cda]::run_compiler::{closure#1}>::{closure#0}, core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>>
  23:     0x7fce65bd021f - std[54a224735ab911a]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[ee304f5e577cf6b9]::util::run_in_thread_pool_with_globals<rustc_interface[ee304f5e577cf6b9]::interface::run_compiler<core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>, rustc_driver[a243c3d9738e9cda]::run_compiler::{closure#1}>::{closure#0}, core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>>::{closure#0}, core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>>
  24:     0x7fce65bd0509 - <<std[54a224735ab911a]::thread::Builder>::spawn_unchecked_<rustc_interface[ee304f5e577cf6b9]::util::run_in_thread_pool_with_globals<rustc_interface[ee304f5e577cf6b9]::interface::run_compiler<core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>, rustc_driver[a243c3d9738e9cda]::run_compiler::{closure#1}>::{closure#0}, core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>>::{closure#0}, core[f66f8263def5172a]::result::Result<(), rustc_errors[e9d593471e7e5a86]::ErrorGuaranteed>>::{closure#1} as core[f66f8263def5172a]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  25:     0x7fce636128c3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hb7ce1f9a74daab59
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/alloc/src/boxed.rs:1854:9
  26:     0x7fce636128c3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc28d1f774d95b80d
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/alloc/src/boxed.rs:1854:9
  27:     0x7fce636128c3 - std::sys::unix::thread::Thread::new::thread_start::h7c8916b17d3c6d45
                               at /rustc/c274e4969f058b1c644243181ece9f829efa7594/library/std/src/sys/unix/thread.rs:108:17
  28:     0x7fce63542609 - start_thread
  29:     0x7fce6345b163 - clone
  30:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.61.0-nightly (c274e4969 2022-03-05) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
warning: `playground` (bin "playground") generated 2 warnings
error: could not compile `playground`; 2 warnings emitted
```

### LINKED ISSUE #94999: ICE on stable when trying to imitate `trivial_bounds`
Trying to emulate trivial bounds via generics & associated types panics on stable, beta and nightly.

### Code

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7fd4f152df6e03d96a0f4e6c0b807cfd

```Rust
trait Identity<Q> {
    type T;
}

impl<Q, T> Identity<Q> for T {
    type T = T;
}

trait Holds {
    type Q;
}


struct S;
struct X(S);

struct XHelper;

impl Holds for X {
    type Q = XHelper;
}

impl<Q> Clone for X where <S as Identity<Q>>::T: Clone, X: Holds<Q = Q> {
    fn clone(&self) -> Self {
        Self(self.0.clone())
    }
}
```

### Meta

`rustc --version --verbose`:
```
note: rustc 1.59.0 (9d1b2106e 2022-02-23) running on x86_64-unknown-linux-gnu
```

Reproducible on beta and nightly

### Error output

```
error: internal compiler error: Encountered error `Unimplemented` selecting `Binder(<S as std::clone::Clone>, [])` during codegen
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/codegen.rs:68:32

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1188:13
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:     0x7ff60cfa58ad - std::backtrace_rs::backtrace::libunwind::trace::ha8f287310bdbcd77
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7ff60cfa58ad - std::backtrace_rs::backtrace::trace_unsynchronized::h7623a82dbfb4fb46
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7ff60cfa58ad - std::sys_common::backtrace::_print_fmt::hf1e6d43252933384
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7ff60cfa58ad - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf9b8d0350fa16147
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7ff60d00552c - core::fmt::write::hc4f5a0569d40abb2
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/core/src/fmt/mod.rs:1190:17
   5:     0x7ff60cf95ad8 - std::io::Write::write_fmt::h61cc67cfa4ed049d
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/io/mod.rs:1657:15
   6:     0x7ff60cfa9887 - std::sys_common::backtrace::_print::hb756cfd40b0abea2
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7ff60cfa9887 - std::sys_common::backtrace::print::h18bde2ddc5524da3
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7ff60cfa9887 - std::panicking::default_hook::{{closure}}::he75def497ad30ae6
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/panicking.rs:295:22
   9:     0x7ff60cfa9550 - std::panicking::default_hook::h0635b9c9300448b4
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/panicking.rs:314:9
  10:     0x7ff60d735731 - rustc_driver[fde845658e6d28a1]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7ff60cfaa164 - std::panicking::rust_panic_with_hook::hc3f0194fa4cee89d
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/panicking.rs:702:17
  12:     0x7ff60cfa9e17 - std::panicking::begin_panic_handler::{{closure}}::h9a433f193b926a3c
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/panicking.rs:588:13
  13:     0x7ff60cfa5d74 - std::sys_common::backtrace::__rust_end_short_backtrace::h35fa457aa178928e
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7ff60cfa9b29 - rust_begin_unwind
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/panicking.rs:584:5
  15:     0x7ff60cf728a3 - core::panicking::panic_fmt::h2442b4ae872c60ae
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/core/src/panicking.rs:143:14
  16:     0x7ff60e821a7c - core[a1225df82ec0a19e]::panicking::panic_display::<&str>
  17:     0x7ff60fe7695d - <rustc_errors[2de1ab05ff8f630e]::HandlerInner>::flush_delayed
  18:     0x7ff60fe74dba - <rustc_errors[2de1ab05ff8f630e]::HandlerInner as core[a1225df82ec0a19e]::ops::drop::Drop>::drop
  19:     0x7ff60f5990f6 - core[a1225df82ec0a19e]::ptr::drop_in_place::<rustc_session[9bb0fc831dcf3ae9]::parse::ParseSess>
  20:     0x7ff60f59b99a - <alloc[c470e77ec7ef658d]::rc::Rc<rustc_session[9bb0fc831dcf3ae9]::session::Session> as core[a1225df82ec0a19e]::ops::drop::Drop>::drop
  21:     0x7ff60f5807bc - core[a1225df82ec0a19e]::ptr::drop_in_place::<rustc_interface[8e82cc6f61f52ee6]::interface::Compiler>
  22:     0x7ff60f58945e - rustc_interface[8e82cc6f61f52ee6]::interface::create_compiler_and_run::<core[a1225df82ec0a19e]::result::Result<(), rustc_errors[2de1ab05ff8f630e]::ErrorReported>, rustc_driver[fde845658e6d28a1]::run_compiler::{closure#1}>
  23:     0x7ff60f56e6a4 - std[eb402c9241147050]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[8e82cc6f61f52ee6]::util::run_in_thread_pool_with_globals<rustc_interface[8e82cc6f61f52ee6]::interface::run_compiler<core[a1225df82ec0a19e]::result::Result<(), rustc_errors[2de1ab05ff8f630e]::ErrorReported>, rustc_driver[fde845658e6d28a1]::run_compiler::{closure#1}>::{closure#0}, core[a1225df82ec0a19e]::result::Result<(), rustc_errors[2de1ab05ff8f630e]::ErrorReported>>::{closure#0}, core[a1225df82ec0a19e]::result::Result<(), rustc_errors[2de1ab05ff8f630e]::ErrorReported>>
  24:     0x7ff60f5a1929 - <<std[eb402c9241147050]::thread::Builder>::spawn_unchecked_<rustc_interface[8e82cc6f61f52ee6]::util::run_in_thread_pool_with_globals<rustc_interface[8e82cc6f61f52ee6]::interface::run_compiler<core[a1225df82ec0a19e]::result::Result<(), rustc_errors[2de1ab05ff8f630e]::ErrorReported>, rustc_driver[fde845658e6d28a1]::run_compiler::{closure#1}>::{closure#0}, core[a1225df82ec0a19e]::result::Result<(), rustc_errors[2de1ab05ff8f630e]::ErrorReported>>::{closure#0}, core[a1225df82ec0a19e]::result::Result<(), rustc_errors[2de1ab05ff8f630e]::ErrorReported>>::{closure#1} as core[a1225df82ec0a19e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  25:     0x7ff60cfb5d43 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h95a31efa3bb13b59
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/alloc/src/boxed.rs:1854:9
  26:     0x7ff60cfb5d43 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h65d3a2f2433c608a
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/alloc/src/boxed.rs:1854:9
  27:     0x7ff60cfb5d43 - std::sys::unix::thread::Thread::new::thread_start::h28cc0caaa86ab15b
                               at /rustc/99f967e7ed97803517e5fad18d27698cab977808/library/std/src/sys/unix/thread.rs:108:17
  28:     0x7ff60cee9609 - start_thread
  29:     0x7ff60ce02163 - clone
  30:                0x0 - <unknown>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by compiler-errors**
Fix a few cases where impossible founds can be found and can cause ICEs in codegen/linting.

Fixes #94999 
Fixes #94680
Actually fixes #93008 this time

### CODE REVIEW

File: compiler/rustc_middle/src/query/mod.rs [hunk 1]
```diff
@@ -1871,8 +1871,40 @@ rustc_queries! {
...
+    // (e.g. we are evaluating a function without substituting generic
+    // parameters), so this filtering serves two purposes:
+    //
+    // 1. We skip evaluating any predicates that we would
+    // never be able prove are unsatisfiable (e.g. `<T as Foo>`
+    // 2. We avoid trying to normalize predicates involving generic
+    // parameters (e.g. `<T as Foo>::MyItem`). This can confuse
+    // the normalization code (leading to cycle errors), since
+    // it's usually never invoked in this way.
+    query check_impossible_predicates_for_item(key: DefId) -> bool {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/query/mod.rs:1904]: Could you rename the query to make the `bool` more explicit?
Something like `item_has_impossible_predicates`?

File: compiler/rustc_middle/src/query/mod.rs [hunk 2]
```diff
@@ -1871,8 +1871,40 @@ rustc_queries! {
...
+        }
+    }
+
+    // Check if it's even possible to satisfy the 'where' clauses
+    // for this item, without substitutions.
+    //
+    // This branch will never be taken for any normal function.
+    // However, it's possible to `#!feature(trivial_bounds)]` to write
+    // a function with impossible to satisfy clauses, e.g.:
+    // `fn foo() where String: Copy {}`
```
Comment (cjgillot) [inline compiler/rustc_middle/src/query/mod.rs:1885]: This paragraph is outdated.

**Review by cjgillot — COMMENTED** [review]
The query you created makes it seem like having impossible predicates is a property of an item. Should this check be done earlier, ie. at the beginning of typeck, in order to skip all the work of type checking and MIR building?

**Comment by compiler-errors** [comment]
@cjgillot:

> The query you created makes it seem like having impossible predicates is a property of an item. Should this check be done earlier, ie. at the beginning of typeck, in order to skip all the work of type checking and MIR building?

Yes, I agree that it's a property of the where clause on an item. However, they are allowed on the item with `feature(trivial_bounds)` and even in some [cases without that feature](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=026b56074fe7334f816ce9b315498414) (which, as a side note, I am probably going to put up a PR to warn on the latter cases).

The current behavior is to typecheck the items as if these predicates are true. I think we should at least continue to typecheck the items given the false predicates -- though I wouldn't be opposed to skipping mir-build (or at least all of mir-opt) in the case that they have impossible predicates. 

This PR doesn't currently do that because I wanted to keep the affected compiler surface small, since it's the inliner that I have seen with the most ICEs due to calling `codegen_fulfill_obligation` on impossible predicates. Do you mind if I investigate additional changes (e.g. to skip all of MIR) in a follow-up PR?

**Comment by compiler-errors** [comment]
@rustbot ready

r? rust-lang/compiler

File: compiler/rustc_trait_selection/src/traits/mod.rs [hunk 3]
```diff
@@ -454,17 +454,38 @@ pub fn impossible_predicates<'tcx>(
...
+fn instantiated_item_has_impossible_predicates<'tcx>(
     tcx: TyCtxt<'tcx>,
     key: (DefId, SubstsRef<'tcx>),
 ) -> bool {
-    debug!("subst_and_check_impossible_predicates(key={:?})", key);
+    debug!("instantiated_item_has_impossible_predicates(key={:?})", key);
 
     let mut predicates = tcx.predicates_of(key.0).instantiate(tcx, key.1).predicates;
     predicates.retain(|predicate| !predicate.needs_subst());
     let result = impossible_predicates(tcx, predicates);
```
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/mod.rs:465]: It's surprising to me that this doesn't just delegate to `item_has_impossible_predicates_for_item`.

The behavior these two are slightly different; could they be merged?

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/mod.rs:465]: I tried to before, but I had some regressions. 

I think I could try again if I'm smarter with some things though... I'll investigate this weekend.

Comment (jackh726) [reply compiler/rustc_trait_selection/src/traits/mod.rs:465]: Do you think we should land this in the meantime? If so, I think I'd like a FIXME added, but other than that, r=me

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/mod.rs:465]: It's not ridiculously critical, so I'll play around with it this weekend and either r=you with a better description of why they're different, or unify these two.

**Comment by compiler-errors** [comment]
Oof, that perf regression sucks. It's possibly because I'm [elaborating the predicates always](https://github.com/rust-lang/rust/pull/95398/files#diff-b81e352311722abf2e93dff8b5af345f3a434adf9c7c068b17de9c815ec114d7R432).

Testing the previous implementation for perf regression in #96382.

**Comment by compiler-errors** [comment]
This was superseded by #96806

**Pull request closed without merging**